packages/core/src/i18n/detect-language.ts

```diff
@@ -0,0 +1,53 @@
+import { IncomingHttpHeaders } from 'http';
+
+import { Optional } from '@logto/essentials';
+import { ParameterizedContext } from 'koa';
+import { IRouterParamContext } from 'koa-router';
+
+/**
+ * Resolve language and its q value from string.
+ * @param languageString The language string in header, e.g. 'en-GB;q=0.8', 'zh-CN'
+ * @returns `[language, q]`, e.g. `['en-GB', 0.8]`; `undefined` if no language is detected.
+ */
+const resolveLanguage = (languageString: string): Optional<[string, number]> => {
+  // Edited from https://github.com/lxzxl/koa-i18next-detector/blob/master/src/lookups/header.js
+  const [language, ...rest] = languageString.split(';');
+
+  if (!language) {
+    return;
+  }
+
+  for (const item of rest) {
+    const [key, value] = item.split('=');
+    if (key === 'q' && !Number.isNaN(Number(value))) {
+      return [language, Number(value)];
+    }
+  }
+
+  return [language, 1];
+};
+
+const normalizeValueToStringArray = (value?: string | string[]): string[] => {
```

Comment:
nit: do we need a util place to keep these general formatting method?

Reply:
yes, this will be better to move into `essentials` packages. added in [LOG-79](https://linear.app/silverhand/issue/LOG-79/update-essentials-packge).